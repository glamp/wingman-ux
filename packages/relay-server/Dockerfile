# Multi-stage Docker build for Wingman Unified Server

# Build stage
FROM node:20-alpine AS builder

WORKDIR /app

# Copy workspace files
COPY package*.json ./
COPY packages/relay-server/package*.json ./packages/relay-server/
COPY packages/shared/package*.json ./packages/shared/

# Install dependencies
RUN npm ci --workspace=@wingman/relay-server --workspace=@wingman/shared

# Copy source code
COPY packages/relay-server ./packages/relay-server
COPY packages/shared ./packages/shared

# Build TypeScript
RUN npm run build --workspace=@wingman/shared
RUN npm run build --workspace=@wingman/relay-server

# Copy pre-built webapp dist (built locally before deployment)
COPY packages/webapp/dist ./packages/webapp/dist

# Runtime stage
FROM node:20-alpine

WORKDIR /app

# Install production dependencies only
COPY package*.json ./
COPY packages/relay-server/package*.json ./packages/relay-server/
COPY packages/shared/package*.json ./packages/shared/

RUN npm ci --workspace=@wingman/relay-server --workspace=@wingman/shared --omit=dev

# Copy built application
COPY --from=builder /app/packages/relay-server/dist ./packages/relay-server/dist
COPY --from=builder /app/packages/shared/dist ./packages/shared/dist
COPY --from=builder /app/packages/webapp/dist ./packages/webapp/dist

# Create wingman storage directory
RUN mkdir -p /data/.wingman/annotations /data/.wingman/sessions

# Set environment variables
ENV NODE_ENV=production
ENV PORT=8787
ENV HOST=0.0.0.0
ENV STORAGE_PATH=/data/.wingman

# Health check
HEALTHCHECK --interval=30s --timeout=3s --start-period=5s --retries=3 \
  CMD node -e "require('http').get('http://localhost:8787/health', (res) => process.exit(res.statusCode === 200 ? 0 : 1))"

# Expose port
EXPOSE 8787

# Start server
CMD ["node", "packages/relay-server/dist/index.js"]