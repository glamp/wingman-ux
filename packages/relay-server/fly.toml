# fly.toml app configuration for Wingman Unified Server (Cloud Mode)
# Deploy with: fly deploy --config packages/relay-server/fly.toml

app = 'wingman-tunnel'
primary_region = 'iad'
kill_signal = 'SIGTERM'
kill_timeout = '30s'

[experimental]
  auto_rollback = true

[build]

[env]
  NODE_ENV = 'production'
  PORT = '8787'
  HOST = '0.0.0.0'
  # Enable cloud tunnel mode
  TUNNEL_MODE = 'cloud'
  # Base URL for generating tunnel URLs
  TUNNEL_BASE_URL = 'wingmanux.com'

[[services]]
  protocol = 'tcp'
  internal_port = 8787
  processes = ['app']

  [[services.ports]]
    port = 80
    handlers = ['http']
    force_https = true

  [[services.ports]]
    port = 443
    handlers = ['tls', 'http']

  [services.concurrency]
    type = 'connections'
    hard_limit = 100
    soft_limit = 80

  [[services.tcp_checks]]
    interval = '10s'
    timeout = '2s'
    grace_period = '30s'

  [[services.http_checks]]
    interval = '30s'
    timeout = '3s'
    grace_period = '5s'
    method = 'GET'
    path = '/health'
    protocol = 'http'
    tls_skip_verify = false

    [services.http_checks.headers]
      Accept = 'application/json'

[[vm]]
  memory = '2gb'
  cpu_kind = 'shared'
  cpus = 2