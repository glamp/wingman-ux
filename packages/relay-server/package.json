{
  "name": "@wingman/relay-server",
  "version": "1.0.0",
  "private": true,
  "main": "dist/index.js",
  "scripts": {
    "build": "tsc",
    "dev": "tsx watch src/index.ts",
    "start": "node dist/index.js",
    "clean": "rm -rf dist",
    "test": "vitest run",
    "test:watch": "vitest watch",
    "test:coverage": "vitest run --coverage",
    "coverage": "npm run test:coverage"
  },
  "dependencies": {
    "@modelcontextprotocol/sdk": "^1.17.4",
    "@roamhq/wrtc": "^0.9.0",
    "@wingman/shared": "*",
    "canvas": "^3.2.0",
    "cors": "^2.8.5",
    "express": "^4.21.2",
    "express-rate-limit": "^7.5.0",
    "http-proxy-middleware": "^3.0.5",
    "simple-peer": "^9.11.1",
    "zod": "^3.25.76"
  },
  "devDependencies": {
    "@types/cors": "^2.8.17",
    "@types/express": "^5.0.0",
    "@types/node": "^22.12.0",
    "tsx": "^4.19.2",
    "typescript": "^5.7.3"
  }
}
