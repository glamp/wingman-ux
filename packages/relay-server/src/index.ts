import express from 'express';
import rateLimit from 'express-rate-limit';
import type { Server } from 'http';
import { createProxyMiddleware } from 'http-proxy-middleware';
import path from 'path';
import { errorHandler } from './middleware/error-handler';
import { createCorsMiddleware, corsDebugMiddleware } from './middleware/cors';
import { annotationsRouter } from './routes/annotations';
import { healthRouter } from './routes/health';
import { mcpRouter } from './routes/mcp';
import { tunnelRouter } from './routes/tunnel';
import { StorageService } from './services/storage';
import { createLogger } from '@wingman/shared';

export interface ServerOptions {
  port?: number;
  host?: string;
  storagePath?: string;
}

const logger = createLogger('Wingman:RelayServer');

export function createServer(options: ServerOptions = {}) {
  const app = express();
  const port = options.port ?? 8787;
  const host = options.host || 'localhost';
  const storagePath = options.storagePath || './wingman/annotations';

  // Initialize storage
  const storage = new StorageService(storagePath);

  // CORS Middleware - configured for browser extensions
  app.use(corsDebugMiddleware); // Optional: logs CORS requests when DEBUG_CORS=true
  app.use(createCorsMiddleware());
  app.use(express.json({ limit: '25mb' }));
  app.use(express.urlencoded({ extended: true, limit: '25mb' }));

  // Rate limiting
  const limiter = rateLimit({
    windowMs: 1 * 60 * 1000, // 1 minute
    max: 100, // 100 requests per minute
  });
  app.use('/annotations', limiter);

  // Routes
  app.use('/health', healthRouter);
  app.use('/annotations', annotationsRouter(storage));
  app.use('/mcp', mcpRouter(storage));
  app.use('/tunnel', tunnelRouter());

  logger.debug('NODE_ENV:', process.env.NODE_ENV);

  // Serve preview UI - proxy to dev server in development, static files in production
  // Skip preview UI setup during testing to avoid proxy issues
  if (process.env.NODE_ENV !== 'test' && process.env.NODE_ENV === 'development') {
    logger.info('Using development mode - proxying to preview UI dev server');
    // In development, proxy to the preview UI dev server
    app.use(
      '/preview',
      createProxyMiddleware({
        target: 'http://localhost:3001/preview',
        changeOrigin: true,
        ws: true,
        pathRewrite: {
          '^/preview': '', // Remove /preview prefix since target already includes it
        },
        logger: console,
        on: {
          proxyReq: (proxyReq, req) => {
            logger.debug('Proxying:', req.method, req.url, '->', proxyReq.path);
          },
        },
      })
    );
  } else {
    // In production, serve static files
    const previewUIPath = path.resolve(process.cwd(), '../preview-ui/dist');
    app.use(
      '/preview',
      express.static(previewUIPath, {
        index: 'index.html',
        fallthrough: false,
      })
    );

    // Handle preview UI routes - serve index.html for SPA routing
    app.get('/preview/*', (_req, res) => {
      res.sendFile(path.join(previewUIPath, 'index.html'));
    });
  }

  // Error handler
  app.use(errorHandler);

  // Start server function
  const start = (): Promise<Server> => {
    return new Promise((resolve, reject) => {
      const server = app
        .listen(port, host, () => {
          const address = server.address();
          const actualPort =
            typeof address === 'string' ? parseInt(address) : address?.port || port;
          logger.info(`ðŸª¶ Wingman Relay Server running on http://${host}:${actualPort}`);
          logger.info(`Health check: http://${host}:${actualPort}/health`);
          logger.info(`Annotations endpoint: http://${host}:${actualPort}/annotations`);
          logger.info(`ðŸª¶ MCP endpoint: http://${host}:${actualPort}/mcp (for Claude Code)`);
          resolve(server);
        })
        .on('error', reject);
    });
  };

  return { app, start };
}

// If running directly (not imported)
if (require.main === module) {
  const PORT = process.env.PORT ? parseInt(process.env.PORT, 10) : 8787;
  const HOST = process.env.HOST || 'localhost';
  
  let server: Server | null = null;

  // Graceful shutdown handlers
  const shutdown = async (signal: string) => {
    logger.info(`${signal} received, shutting down gracefully...`);
    if (server) {
      server.close(() => {
        logger.info('Server closed');
        process.exit(0);
      });
      
      // Force exit after 5 seconds
      setTimeout(() => {
        logger.error('Could not close connections in time, forcefully shutting down');
        process.exit(1);
      }, 5000);
    } else {
      process.exit(0);
    }
  };

  // Register shutdown handlers
  process.on('SIGTERM', () => shutdown('SIGTERM'));
  process.on('SIGINT', () => shutdown('SIGINT'));

  createServer({ port: PORT, host: HOST })
    .start()
    .then((s) => {
      server = s;
    })
    .catch((error) => {
      logger.error('Failed to start server:', error);
      process.exit(1);
    });
}
