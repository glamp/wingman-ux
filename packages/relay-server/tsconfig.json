{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "rootDir": "./src",
    "outDir": "./dist",
    "lib": ["es2020"],
    "target": "es2020",
    "module": "commonjs",
    "types": ["node"]
  },
  "include": ["src/**/*"],
  "exclude": ["node_modules", "dist", "src/**/*.test.ts", "src/**/*.spec.ts"],
  "references": [
    { "path": "../shared" }
  ]
}