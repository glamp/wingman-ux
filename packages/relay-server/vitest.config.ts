import { defineConfig } from 'vitest/config';
import path from 'path';

export default defineConfig({
  test: {
    globals: true,
    environment: 'node',
    setupFiles: [],
    env: {
      NODE_ENV: 'test'
    },
    coverage: {
      provider: 'v8',
      reporter: ['text', 'json', 'html', 'json-summary'],
      reportsDirectory: './coverage',
      exclude: [
        'node_modules/**',
        'dist/**',
        '**/*.config.ts',
        '**/*.test.ts',
        '**/__tests__/**',
        '**/types.ts'
      ],
      thresholds: {
        statements: 50,
        branches: 50,
        functions: 50,
        lines: 50
      },
      all: true,
      clean: true
    }
  },
  resolve: {
    alias: {
      '@': path.resolve(__dirname, './src')
    }
  }
});