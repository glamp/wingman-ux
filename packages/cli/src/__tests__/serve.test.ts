import { describe, it, expect, afterEach } from 'vitest';
import { spawn, ChildProcess } from 'child_process';
import { createServer } from '@wingman/relay-server';
import type { Server } from 'http';
import path from 'path';
import { fileURLToPath } from 'url';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const CLI_PATH = path.resolve(__dirname, '../../dist/index.js');

describe('wingman serve command', () => {
  let server: Server | null = null;
  let cliProcess: ChildProcess | null = null;

  afterEach(async () => {
    // Clean up any running servers
    if (server) {
      await new Promise<void>((resolve) => {
        server!.close(() => resolve());
      });
      server = null;
    }

    // Clean up any running CLI processes
    if (cliProcess) {
      cliProcess.kill();
      cliProcess = null;
    }
  });

  it('should display help for serve command', async () => {
    const output = await runCliCommand(['serve', '--help']);
    expect(output).toMatchSnapshot();
    expect(output).toContain('Start the Wingman relay server');
    expect(output).toContain('-p, --port <number>');
    expect(output).toContain('-h, --host <address>');
  });

  it('should start server on default port', async () => {
    const { start } = createServer({ port: 8787, host: 'localhost' });
    server = await start();
    
    const address = server.address();
    expect(address).toBeTruthy();
    if (address && typeof address !== 'string') {
      expect(address.port).toBe(8787);
    }
  });

  it('should start server on custom port', async () => {
    const { start } = createServer({ port: 9999, host: 'localhost' });
    server = await start();
    
    const address = server.address();
    expect(address).toBeTruthy();
    if (address && typeof address !== 'string') {
      expect(address.port).toBe(9999);
    }
  });

  it('should bind to custom host', async () => {
    const { start } = createServer({ port: 8788, host: '127.0.0.1' });
    server = await start();
    
    const address = server.address();
    expect(address).toBeTruthy();
    if (address && typeof address !== 'string') {
      expect(address.address).toBe('127.0.0.1');
    }
  });

  it('should handle port already in use', async () => {
    // Start a server on port 8789
    const { start } = createServer({ port: 8789, host: 'localhost' });
    server = await start();

    // Try to start another server on the same port
    try {
      const { start: start2 } = createServer({ port: 8789, host: 'localhost' });
      await start2();
      throw new Error('Should have thrown EADDRINUSE');
    } catch (error: any) {
      expect(error.code).toBe('EADDRINUSE');
    }
  });

  it('should validate port range', async () => {
    const invalidPorts = [0, -1, 65536, 100000, NaN];
    
    for (const port of invalidPorts) {
      if (isNaN(port) || port < 1 || port > 65535) {
        // This is the validation we expect in the CLI
        expect(port).toSatisfy((p: number) => isNaN(p) || p < 1 || p > 65535);
      }
    }
  });
});

// Helper function to run CLI commands
function runCliCommand(args: string[]): Promise<string> {
  return new Promise((resolve, reject) => {
    const child = spawn('node', [CLI_PATH, ...args], {
      env: { ...process.env },
    });

    let output = '';
    let errorOutput = '';

    child.stdout.on('data', (data) => {
      output += data.toString();
    });

    child.stderr.on('data', (data) => {
      errorOutput += data.toString();
    });

    child.on('close', (code) => {
      if (code === 0 || args.includes('--help')) {
        resolve(output);
      } else {
        reject(new Error(`CLI exited with code ${code}: ${errorOutput}`));
      }
    });

    child.on('error', reject);
  });
}