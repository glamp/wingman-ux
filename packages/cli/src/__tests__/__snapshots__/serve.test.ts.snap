// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`wingman serve command > should display help for serve command 1`] = `
"Usage: wingman serve [options]

Start the Wingman relay server

Options:
  -p, --port <number>   port to listen on (default: "8787")
  -h, --host <address>  host/address to bind to (default: "localhost")
  --help                display help for command
"
`;
